src/types/impl-trait.md

```diff
@@ -5,33 +5,120 @@
 >
 > _ImplTraitTypeOneBound_ : `impl` [_TraitBound_]
 
-## Anonymous type parameters
+> **Edition differences**: `impl Trait` is new in the 2018 edition.
+
+`impl Trait` provides ways to specify unnamed but concrete types that
+implement a specific trait.
+It can appear in two sorts of places: argument position (where it can act as an anonymous type parameter to functions), and return position (where it can act as an abstract return type).
 
-> Note: This section is a placeholder for more comprehensive reference
-> material.
+```rust
+trait Trait {}
+# impl Trait for () {}
+
+// argument position: anonymous type parameter
+fn foo(arg: impl Trait) {
+}
+
+// return position: abstract return type
+fn bar() -> impl Trait {
+}
+```
+## Anonymous type parameters
 
 > Note: This is often called "impl Trait in argument position".
 
-Functions can declare an argument to be an anonymous type parameter where the
-callee must provide a type that has the bounds declared by the anonymous type
-parameter and the function can only use the methods available by the trait
-bounds of the anonymous type parameter.
+Functions can use `impl` followed by a set of trait bounds to declare an argument as having an anonymous type.
+The caller must provide a type that satisfies the bounds declared by the anonymous type parameter, and the function can only use the methods available through the trait bounds of the anonymous type parameter.
 
-They are written as `impl` followed by a set of trait bounds.
+For example, these two forms are almost equivalent:
 
-## Abstract return types
+```rust,ignore
+trait Trait {}
+
+// generic type parameter
+fn foo<T: Trait>(arg: T) {
+}
 
-> Note: This section is a placeholder for more comprehensive reference
-> material.
+// impl Trait in argument position
+fn foo(arg: impl Trait) {
+}
+```
+
+That is, `impl Trait` in argument position is syntactic sugar for a generic type parameter like `<T: Trait>`, except that the type is anonymous and doesn't appear in the [_GenericParams_] list.
+
+> **Note:**
+> For function arguments, generic type parameters and `impl Trait` are not exactly equivalent.
+> With a generic parameter such as `<T: Trait>`, the caller has the option to explicitly specify the generic argument for `T` at the call site using [_GenericArgs_], for example, `foo::<usize>(1)`.
+> If `impl Trait` is the type of a function argument, then the caller can't ever specify the type of that argument by using a generic argument.
+>
+> Therefore, changing the function signature from either one to the other can constitute a breaking change for the callers of a function.
```

Comment:
I can see how changing from generic parameter to impl trait would be a breaking change, but can you explain how the reverse is also a breaking change?

Reply:
I think this was based on text that I copied from the edition guide. For the reverse direction, I'm not especially familiar with using turbofish when there are multiple type parameters, e.g., what if we went from

```rust
fn foo<T: PartialOrd>(a: T, b: impl PartialOrd) {}
foo::<i32>(0, 1.0);
```

to

```rust
fn foo<T: PartialOrd, U: PartialOrd>(a: T, b: U) {}
foo::<i32>(0, 1.0);
```

Is it valid to omit the second type parameter? How does the compiler know which one is omitted? (I can't find any text in the reference about this case.) I'm sorry if this is a bad example.

Reply:
I experimented a bit in the playground, and it seems like in practice:
* If there are multiple type parameters, either all or none must be specified by turbofish
* If there is a mixture of type parameters and `impl Trait` in argument position, turbofish may not be used for any type arguments at all

Side note: are we avoiding using the term "turbofish" in the reference?

Reply:
I see.  The second parameter can be omitted if it is defaulted, like `<T, U=T>`, but that can't be done with functions. 

The reference normally doesn't discuss semver compatibility since that is a cargo concept (documented [here](https://doc.rust-lang.org/nightly/cargo/reference/semver.html)). It's probably fine to leave this as-is, though.

There is one mention of turbofish, plus another in the glossary, but otherwise it usually doesn't come up.

Reply:
Some further experimentation suggests that both my proposed text and the edition guide are differently wrong:
* `impl Trait` in return position doesn't prevent the use of turbofish, contrary to the edition guide, which says that "In the case of impl Trait, if it is used anywhere in the function definition, then you can't use turbo-fish at all"
* `impl Trait` in any parameter type prevents the caller from providing any generic arguments at all, including return type and even const generics (rust-lang/rust#85475)! This is contrary to my text, which says it only prevents turbofish for that parameter.